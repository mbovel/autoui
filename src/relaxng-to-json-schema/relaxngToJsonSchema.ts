import { ensure } from "../model/utils";
import { JSONSchema7 } from "json-schema";

const AUNS = "https://github.com/mbovel/autoui";
const RGNG = "http://relaxng.org/ns/structure/1.0";

export function relaxngToJsonSchema(schemaString: string): JSONSchema7 {
	const fullSchema = parseXML(schemaString);
	return convert(fullSchema.children[0], fullSchema)[1];
}

let noNameCounter = 1;

function convert(el: Element, root: Document): [string, JSONSchema7] {
	let schema: JSONSchema7;
	switch (el.tagName) {
		case "zeroOrMore":
			schema = { type: "array", items: convertChildren(el.children, root) };
			break;
		case "oneOrMore":
			schema = { type: "array", items: convertChildren(el.children, root), minItems: 1 };
			break;
		case "optional":
			return convert(el.children[0], root);
		case "element":
		case "attribute":
		case "group":
			schema = convertChildren(el.children, root);
			break;
		case "choice":
			const choices = [...el.children].map(c => convert(c, root)[1]);
			schema = {
				properties: { "autoui:choice": { type: "number", enum: [...choices.keys()] } },
				oneOf: choices
			};
			break;
		case "grammar":
			schema = convertChildren(find(root, "start").children, root);
			break;
		case "value":
			schema = { type: "string", const: el.textContent };
			break;
		case "data":
			schema = { type: "string" };
			break;
		case "text":
			schema = { type: "string" };
			break;
		case "ref":
			const name = el.getAttribute("name");
			ensure(name, "Missing name attribute");
			const defineEl = find(root, `define[name='${name}']`);
			schema = convertChildren(defineEl.children, root);
			break;
		default:
			throw new Error("Unknown tag name: " + el.tagName);
	}
	const key =
		el.getAttribute("name") ?? el.getAttributeNS(AUNS, "key") ?? `noName${++noNameCounter}`;
	schema.title = el.getAttribute("name") ?? el.getAttributeNS(AUNS, "title") ?? undefined;
	return [key, schema];
}

function convertChildren(children: HTMLCollection, root: Document): JSONSchema7 {
	if (children.length === 1) return convert(children[0], root)[1];
	return { type: "object", properties: Object.fromEntries(convertAll(children, root)) };
}

function convertAll(els: HTMLCollection, root: Document) {
	return [...els].filter(isRelevant).map(c => convert(c, root));
}

/*function xpathNsResolver(prefix: string | null): string | null {
	var ns: { [key: string]: string | undefined } = {
		rng: RGNG
	};
	return (prefix && ns[prefix]) ?? null;
}*/

function find(root: Document, query: string) {
	// Does not work in JSDom
	/*const result = root
		.evaluate(xpath, context, xpathNsResolver, XPathResult.UNORDERED_NODE_ITERATOR_TYPE)
		.iterateNext();*/
	const result = root.querySelector(query);
	ensure(result, "XPath query " + query + " returned no result");
	ensure(isElement(result), "XPath query " + query + " returned a non-element node");
	return result;
}

const nodeNames = new Set([
	"zeroOrMore",
	"oneOrMore",
	"optional",
	"element",
	"attribute",
	"group",
	"choice",
	"grammar",
	"value",
	"data",
	"text",
	"ref"
]);

function isRelevant(node: Element) {
	return node.namespaceURI === RGNG && nodeNames.has(node.tagName);
}

function isElement(node: Node): node is Element {
	return node.nodeType === Node.ELEMENT_NODE;
}

function parseXML(string: string) {
	return new DOMParser().parseFromString(string, "text/xml");
}
