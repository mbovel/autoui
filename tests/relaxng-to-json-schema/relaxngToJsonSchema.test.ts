import * as fs from "fs";
import { relaxngToJsonSchema } from "../../src/relaxng-to-json-schema/relaxngToJsonSchema";

const read = (file: string) => fs.readFileSync("tests/relaxng-to-json-schema/" + file, "utf8");

describe("relaxngToJsonSchema", () => {
	for (let i = 1; i <= 6; ++i) {
		test(`correctly converts example ${i}`, () => {
			const result = relaxngToJsonSchema(read(`ex${i}.rng`));
			expect(result).toMatchSnapshot();
		});
	}
});
