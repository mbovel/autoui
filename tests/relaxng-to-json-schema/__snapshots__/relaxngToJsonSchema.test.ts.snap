// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relaxngToJsonSchema correctly converts example 1 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "email": Object {
        "title": "email",
        "type": "string",
      },
      "name": Object {
        "title": "name",
        "type": "string",
      },
    },
    "title": "card",
    "type": "object",
  },
  "title": "addressBook",
  "type": "array",
}
`;

exports[`relaxngToJsonSchema correctly converts example 2 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "email": Object {
        "title": "email",
        "type": "string",
      },
      "name": Object {
        "title": "name",
        "type": "string",
      },
    },
    "title": "card",
    "type": "object",
  },
  "minItems": 1,
  "title": "addressBook",
  "type": "array",
}
`;

exports[`relaxngToJsonSchema correctly converts example 3 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "email": Object {
        "title": "email",
        "type": "string",
      },
      "name": Object {
        "title": "name",
        "type": "string",
      },
      "note": Object {
        "title": "note",
        "type": "string",
      },
    },
    "title": "card",
    "type": "object",
  },
  "title": "addressBook",
  "type": "array",
}
`;

exports[`relaxngToJsonSchema correctly converts example 4 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "email": Object {
        "title": "email",
        "type": "string",
      },
      "noName16": Object {
        "oneOf": Array [
          Object {
            "title": "name",
            "type": "string",
          },
          Object {
            "properties": Object {
              "familyName": Object {
                "title": "familyName",
                "type": "string",
              },
              "givenName": Object {
                "title": "givenName",
                "type": "string",
              },
            },
            "title": undefined,
            "type": "object",
          },
        ],
        "properties": Object {
          "autoui:choice": Object {
            "enum": Array [
              0,
              1,
            ],
            "type": "number",
          },
        },
        "title": undefined,
      },
      "note": Object {
        "title": "note",
        "type": "string",
      },
    },
    "title": "card",
    "type": "object",
  },
  "title": "addressBook",
  "type": "array",
}
`;

exports[`relaxngToJsonSchema correctly converts example 5 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "email": Object {
        "title": "email",
        "type": "string",
      },
      "name": Object {
        "title": "name",
        "type": "string",
      },
    },
    "title": "card",
    "type": "object",
  },
  "title": "addressBook",
  "type": "array",
}
`;

exports[`relaxngToJsonSchema correctly converts example 6 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "Email": Object {
        "title": "Email",
        "type": "string",
      },
      "Name": Object {
        "title": "Name",
        "type": "string",
      },
    },
    "title": "Card",
    "type": "object",
  },
  "title": undefined,
  "type": "array",
}
`;
